gloo/transport/tcp/tls/pair.cc

```diff
@@ -0,0 +1,386 @@
+    case Op::SEND_BUFFER:
+      // Done sending data to pinned buffer; trigger completion.
+      rx_.buf->handleRecvCompletion();
+      break;
+    case Op::SEND_UNBOUND_BUFFER:
+      // Remote side is sending data to unbound buffer; trigger completion
+      buf->handleRecvCompletion(rank_);
+      break;
+    case Op::NOTIFY_SEND_READY:
+      // Remote side has pending send operation
+      handleRemotePendingSend(rx_);
+      break;
+    case Op::NOTIFY_RECV_READY:
+      // Remote side has pending recv operation
+      handleRemotePendingRecv(rx_);
+      break;
+  }
+
+  // Reset read operation state.
+  rx_ = Op();
+  return true;
+}
+
+void Pair::handleEvents(int events) {
+  // Try to acquire the pair's lock so the device thread (the thread
+  // that ends up calling handleEvents) can mutate the tx and rx op
+  // fields of this instance. If the lock cannot be acquired that
+  // means some other thread is trying to mutate this pair's state,
+  // which in turn might require calling into (and locking) the
+  // underlying device (for example, when the pair transitions to the
+  // CLOSED state). To avoid deadlocks, attempt to lock the pair and
+  // skip handling the events until the next tick if the lock cannot
+  // be acquired.
+  std::unique_lock<std::mutex> lock(m_, std::try_to_lock);
+  if (!lock) {
+    return;
+  }
+
+  // State must be <= CONNECTED.
+  // If state is CLOSED; this function will NOT be called. Refer to
+  // Pair::changeState and Device::unregisterDescriptor for more info.
+  GLOO_ENFORCE_LE(state_, CONNECTED);
+
+  // Exception must not be set.
+  // If exception is set, state must advance to CLOSED state.
+  GLOO_ENFORCE(ex_ == nullptr);
+
+  if (state_ == CONNECTED) {
+    if (!is_ssl_connected_) {
+      if (!is_client_) {
+        int events = 0;
+        if ((events = handshake())) {
+          device_->registerDescriptor(fd_, events, this);
+        }
+      }
+    } else {
+      if (events & EPOLLOUT) {
+        GLOO_ENFORCE(
+            !tx_.empty(), "tx_ cannot be empty because EPOLLOUT happened");
+        while (!tx_.empty()) {
+          auto& op = tx_.front();
+          if (!write(op)) {
+            // Write did not complete; wait for epoll.
+            break;
+          }
+          // Write completed; remove from queue.
+          tx_.pop_front();
+        }
+        // If there is nothing to transmit; remove EPOLLOUT.
+        if (tx_.empty()) {
+          device_->registerDescriptor(fd_, EPOLLIN, this);
+        }
+      }
+      if (events & EPOLLIN) {
+        while (read()) {
+          // Keep going
+        }
+      }
+    }
+    return;
+  }
+
+  if (state_ == LISTENING) {
+    handleListening();
+    return;
+  }
+
+  if (state_ == CONNECTING) {
+    handleConnecting();
+    return;
+  }
+
+  GLOO_ENFORCE(false, "Unexpected state: ", state_);
+}
+
+void Pair::changeState(Pair::state nextState) noexcept {
+  if (nextState == CLOSED && is_ssl_connected_) {
+    SSL_shutdown(ssl_);
```

Comment:
do we need to check return value of `SSL_shutdown` in case it is not successful? https://www.openssl.org/docs/man1.0.2/man3/SSL_shutdown.html

Reply:
After some investigations I think that the best what I can do is just call `SSL_shutdown(ssl_)` twice